### PR TITLE
Allow specifying a SSH key name for AWS

### PR DESCRIPTION
**Pull request by johnzeringue**
Related to #2309, this allows naming an existing key pair using the
cluster spec field `sshKeyName`.

In our use case, kops can now be used without providing the ability to
create EC2 key pairs.

### CODE REVIEW

File: pkg/model/names.go [hunk 1]
```diff
@@ -116,14 +116,21 @@ func (b *KopsModelContext) LinkToIAMInstanceProfile(ig *kops.InstanceGroup) *aws
 	return &awstasks.IAMInstanceProfile{Name: &name}
 }
 
-// SSHKeyName computes a unique SSH key name, combining the cluster name and the SSH public key fingerprint
+// SSHKeyName computes a unique SSH key name, combining the cluster name and the SSH public key fingerprint.
+// If an SSH key name is provided in the cluster configuration, it will use that instead.
 func (c *KopsModelContext) SSHKeyName() (string, error) {
```
Comment (chrislovecnm) [inline pkg/model/names.go]: Dumb questions, but when we do a delete a cluster do we delete the key?  We have an approach of having shared items, such as https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/awstasks/vpc.go#L43.  We would probably want that as well, and it may help with delete.

Comment (johnzeringue) [inline pkg/model/names.go]: > when we do a delete a cluster do we delete the key?

I don't think we should. I'm not sure what's been done in the past, but I think that things created by kops should be deleted by kops and things not created by kops should not be deleted by kops.

I don't fully understand the current behavior on deletion (the task lifecycle is hard to follow in my IDE), but my intention is not to delete.

Comment (chrislovecnm) [inline pkg/model/names.go]: @johnzeringue yes we should not delete a shared pre-existing ssh key.  Does kops try with your changes?

Comment (chrislovecnm) [inline pkg/model/names.go]: Also, this is just re-using the aws key, not the ssh public key that is installed on the users' machine? Correct?

**Comment by justinsb** [comment]
Code looks good - thanks @johnzeringue 

I had to check, but looking at the existing code we do compute the AWS-style-fingerprint and thus don't rely on the name.  We currently embed the fingerprint in the name to avoid collisions and because we can't change a public key - and nor can we tag SSH public keys.  And that means we don't delete the key as part of cluster deletion.

But ... I'm wondering about the use case here.  I never thought of the ability to create an SSH key as something particularly powerful (but do LMK!)  We have the lifecycle work going on, which will separate out IAM & SSH permissions, and I know some people have wanted to split those further.  But if the motivation here is to minimize the amount of permissions needed, it seems like the IAM permissions are the big one, not the permission to register an AWS public key!

That said I don't think these are blocking issues to the PR, I would just like to better understand the use case.  I was pondering the name - whether we should put this into an SSH section, but we don't have any other SSH fields (other than `SSHAccess`, which would logically better fit into a networking section anyway) - so I think the schema change is fine.

File: pkg/model/names.go (hunk 1 above)
Comment (johnzeringue) [inline pkg/model/names.go]: > @johnzeringue yes we should not delete a shared pre-existing ssh key. Does kops try with your changes?

In our testing, it seems not.

> Also, this is just re-using the aws key, not the ssh public key that is installed on the users' machine? Correct?

Yes

**Comment by johnzeringue** [comment]
@justinsb the use case is separating the ability to provision EC2 instances with the ability to access those instances. There are other ways to do it, but preventing key creation is an easy way to create a strong access barrier.

**Comment by johnzeringue** [comment]
Documentation added

**Comment by johnzeringue** [comment]
@justinsb any more feedback on this? I'm happy to hop on Slack if you want to talk about use case more in depth

**Comment by chrislovecnm** [comment]
You may need a rebase to get e2e passing.

**Comment by johnzeringue** [comment]
@chrislovecnm I tried rebasing, but e2e is still failing. Not quite sure what's wrong from the logs. Any tips?

**Comment by chrislovecnm** [comment]
@johnzeringue we have been having problems with e2e.  Most likely flakey testing not your code.  How have you tested?

/retest

**Comment by johnzeringue** [comment]
@chrislovecnm we created a new cluster in AWS with a preexisting SSH key. We verified that we could SSH into instances using the preexisting key. We destroyed the cluster and saw that it did not delete or attempt to delete the preexisting key.

**Comment by chrislovecnm** [comment]
@justins  we good on this PR?

**Comment by ApsOps** [comment]
I've felt the need for this use-case too. Is this ready to be merged?

@chrislovecnm I think you mentioned someone else instead of @justinsb by mistake 😅

**Comment by chrislovecnm** [comment]
@justinsb we good with this PR?  I may need this as well.

**Comment by chrislovecnm** [comment]
Thanks @ApsOps

**Comment by chrislovecnm** [comment]
I know certain parties are super busy, merging this in, we can iterate if needs be

**Comment by chrislovecnm** [comment]
@johnzeringue you mind filing an issue to add validation that this is aws only? I think, may be wrong, that GCE does not have this feature
